### PR TITLE
DKG skeleton (continued)

### PR DESCRIPTION
**Pull request by zjma**
### Description
This PR includes:
- some DKG type definition.
- DKG epoch manager skeleton
- DKG state manager skeleton
- DKG network tasks

### CODE REVIEW

File: dkg/src/dkg_manager/mod.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+pub struct DKGManager {
+    my_addr: AccountAddress,
+    epoch_state: EpochState,
+    vtxn_pool_write_cli: Arc<vtxn_pool::SingleTopicWriteClient>,
+    agg_node_producer: Arc<dyn AggNodeProducer>,
+    agg_node_tx: Option<aptos_channel::Sender<(), DKGAggNode>>,
+    //TODO: inner state and sk
+}
+
+#[allow(clippy::never_loop)]
```
Comment (grao1991) [inline dkg/src/dkg_manager/mod.rs:26]: nit: put this right before the loop?

File: dkg/src/epoch_manager.rs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    validator_txn::ValidatorTransaction,
+};
+use aptos_validator_transaction_pool as vtxn_pool;
+use futures::StreamExt;
+use futures_channel::oneshot;
+use std::sync::Arc;
+
+#[allow(dead_code)]
+pub struct EpochManager<P: OnChainConfigProvider> {
+    sk: Option<bls12381::PrivateKey>,
```
Comment (grao1991) [inline dkg/src/epoch_manager.rs:35]: comment for variables.

File: dkg/src/epoch_manager.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+                (peer, rpc_request) = network_receivers.rpc_rx.select_next_some() => {
+                    self.process_rpc_request(peer, rpc_request)
+                },
+                msg = self.vtxn_pull_notification_rx_from_pool.select_next_some() => {
+                    self.process_vtxn_pull_notification(msg)
+                }
+            };
+
+            if let Err(e) = handling_result {
+                error!("{}", e);
```
Comment (grao1991) [inline dkg/src/epoch_manager.rs:126]: nit: consider sampling?

File: dkg/src/types.rs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+/// Once DKG starts, a validator should send this message to peers in order to collect DKG transcripts from peers.
+#[derive(Clone, Serialize, Deserialize, CryptoHasher, Debug, PartialEq)]
+pub struct DKGNodeRequest {
+    dealer_epoch: u64,
+}
+
+/// Contains a DKG transcript and some metadata.
+#[derive(Clone, Serialize, Deserialize, CryptoHasher, Debug, PartialEq)]
+pub struct DKGNode {
+    //TODO
```
Comment (grao1991) [inline dkg/src/types.rs:19]: nit: put your name

**Review by grao1991 — APPROVED** [review]

File: dkg/src/epoch_manager.rs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        self.shutdown_current_processor().await;
+        self.start_new_epoch(reconfig_notification.on_chain_configs)
+            .await;
+        Ok(())
+    }
+
+    async fn shutdown_current_processor(&mut self) {
+        if let Some(tx) = self.dkg_manager_close_tx.take() {
+            let (ack_tx, ack_rx) = oneshot::channel();
+            tx.send(ack_tx).unwrap();
```
Comment (danielxiangzl) [inline dkg/src/epoch_manager.rs:203]: unwrap is safe, right? This is the only way to shutdown dkg manager?

File: dkg/src/network.rs [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+                            protocol,
+                        }),
+                    };
+
+                    if let Err(e) = self.rpc_tx.push(peer_id, (peer_id, req)) {
+                        warn!(error = ?e, "aptos channel closed");
+                    };
+                },
+                _ => {
+                    // Ignored. Currently only RPC is used.
```
Comment (danielxiangzl) [inline dkg/src/network.rs:150]: print error log?

File: dkg/src/network_interface.rs [hunk 7]
```diff
@@ -1,17 +1,52 @@
...
+        message: DKGMessage,
+        rpc_timeout: Duration,
+    ) -> Result<DKGMessage, Error> {
+        let peer_network_id = self.get_peer_network_id_for_peer(peer);
+        self.network_client
+            .send_to_peer_rpc(message, rpc_timeout, peer_network_id)
+            .await
+    }
+
+    // TODO: we shouldn't need to expose this. Migrate the code to handle peer and network ids.
```
Comment (danielxiangzl) [inline dkg/src/network_interface.rs:48]: What do you mean?

Comment (zjma) [reply dkg/src/network_interface.rs:48]: it's copied from [here](https://github.com/aptos-labs/aptos-core/blob/63e5b9543c8ba4ad772d53a70f476159f4f07533/consensus/src/network_interface.rs#L173-L175)

File: dkg/src/epoch_manager.rs (hunk 5 above)
Comment (zjma) [reply dkg/src/epoch_manager.rs:203]: i think it's safe. And yes that's the only way today.

**Review by danielxiangzl — APPROVED** [review]

File: dkg/src/dkg_manager/agg_node_producer.rs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+pub trait AggNodeProducer: Send + Sync {
+    fn start_produce(
+        &self,
+        epoch_state: EpochState,
+        dkg_config: DKGConfig,
+        agg_node_tx: Option<Sender<(), DKGAggNode>>,
+    ) -> AbortHandle;
+}
+
+pub struct DummyAggNodeProducer {}
```
[the remark this answers is not in the document]
Comment (zjma) [reply dkg/src/dkg_manager/agg_node_producer.rs:23]: in this pr it's actually not...

File: dkg/src/network.rs (hunk 6 above)
Comment (zjma) [reply dkg/src/network.rs:150]: if we do anything, it may become a DOS attack vector?

**Pull request merged**